src/module.c

```diff
@@ -10846,10 +11014,506 @@ sds genModulesInfoString(sds info) {
+ * 
+ *     RedisModuleString *strval;
+ *     RedisModuleString *getStringConfigCommand(const char *name, void *privdata) {
+ *         return strval;
+ *     }
+ *     
+ *     int setStringConfigCommand(const char *name, RedisModuleString *new, ...) {
+ *        strval = new;
+ *     }
+ *     ...
+ *     RedisModule_RegisterStringConfig(ctx, "string", REDISMODULE_CONFIG_MODIFIABLE, &getStringConfigCommand, &setStringConfigCommand, NULL);
+ * 
+ * If the registration fails, REDISMODULE_ERR is returned and one of the following 
+ * errno is set:
+ * * EINVAL: The provided flags are invalid for the registration.
+ * * EALREADY: The provided configuration name is already used.
+ */
+int RM_RegisterStringConfig(RedisModuleCtx *ctx, const char *name, unsigned int flags, RedisModuleConfigGetStringFunc getfn, RedisModuleConfigSetStringFunc setfn, void *privdata) {
+    RedisModule *module = ctx->module;
+    if (moduleConfigValidityCheck(module, name, flags, REDISMODULE_CONFIG_STRING)) {
+        return REDISMODULE_ERR;
+    }
+    ModuleConfig newStringConfig = createModuleConfig(sdsnew(name), REDISMODULE_CONFIG_STRING, flags, privdata);
+    addCallbacksToConfig(newStringConfig, getfn, setfn, string);
+    addModuleConfigtoList(module, newStringConfig);
+    flags = maskModuleConfigFlags(flags);
+    addModuleConfig(module->name, name, flags, module);
+    return REDISMODULE_OK;
+}
+
+/* Create an enum config that server clients can interact with via the 
+ * `CONFIG SET`, `CONFIG GET`, and `CONFIG REWRITE` commands. See 
+ * RedisModule_RegisterStringConfig for detailed information about configs. */
+int RM_RegisterEnumConfig(RedisModuleCtx *ctx, const char *name, unsigned int flags, const char **enum_values, int num_enum_vals, RedisModuleConfigGetEnumFunc getfn, RedisModuleConfigSetEnumFunc setfn, void *privdata) {
+    RedisModule *module = ctx->module;
+    if (moduleConfigValidityCheck(module, name, flags, REDISMODULE_CONFIG_ENUM)) {
+        return REDISMODULE_ERR;
+    }
+    ModuleConfig newEnumConfig = createModuleConfig(sdsnew(name), REDISMODULE_CONFIG_ENUM, flags, privdata);
+    addCallbacksToConfig(newEnumConfig, getfn, setfn, enum);
+    newEnumConfig.type_specific.enums.enum_values = zmalloc(num_enum_vals * sizeof(char *));
+    newEnumConfig.type_specific.enums.num_enums = num_enum_vals;
+    for (int i = 0; i < num_enum_vals; i++) {
+        newEnumConfig.type_specific.enums.enum_values[i] = zstrdup(enum_values[i]);
+    }
+    addModuleConfigtoList(module, newEnumConfig);
+    flags = maskModuleConfigFlags(flags);
+    addModuleConfig(module->name, name, flags, module);
+    return REDISMODULE_OK;
+}
+
+/* Applies all pending configurations on the module load. This should be called
+ * after all of the configurations have been registered for the module. This API
+ * needs to be called when configurations are provided in either `MODULE LOADEX`
+ * or provided as startup arguments. */
+int RM_ApplyConfigs(RedisModuleCtx *ctx) {
```

Comment:
I am in favor of a single mechanism that enables delivering multiple configuration parameters in a single shot, which can be used at load time or when a multi-parameter set is performed.

Maybe it'd be easier to have a single `RedisModuleConfigSetFunc` that takes a single struct or array of structs expressing the desired config change to support that, instead of a single call per parameter.

This could also make it easier for a module to perform a `RedisModuleLoadConfigs` at load time, and receive the same array of structs that would be processed by the set callback, making it possible to abort the load if the configuration is invalid.

Reply:
i think such a mechanism will be more complicated to use on the module's side.
i.e. on simple / single configs, the module can just implement a set of callbacks, and doesn't need to worry about anything else.
the apply function and config grouping is something more advanced and rare.

Reply:
One other use case is being able to do initialization based on configs. Internally we have example where we have a datastructure that isn't feasible to change at runtime, so we want initialize it at startup. It has a default size, but this default can be overridden by a config. We need this overridden value available during onLoad(), since ideally we initialize there. We would need some type of deferred loading otherwise. 

The intention of the Apply was just to make it coherent when the configs were now available.

Reply:
@oranagra I'm not bought on that specific option, but I do think the following is a must:

* Have a single mechanism that covers both simple and more advanced use cases.
* Make all configs available at `OnLoad` time, before it returns, so it's possible to abort loading and handle initialization.
* Support atomic multi-parameter set.